http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java

```diff
@@ -130,13 +130,19 @@ private NettyUtils() {
      *
      * @param eventExecutor Executor to run task in.
      * @param runnable Task to run.
+     *
+     * @return The {@code Future} from from the executor.
      */
-    public static void doInEventLoop(EventExecutor eventExecutor, Runnable runnable) {
+    public static Future<?> doInEventLoop(EventExecutor eventExecutor, Runnable runnable) {
         if (eventExecutor.inEventLoop()) {
-            runnable.run();
-        } else {
-            eventExecutor.submit(runnable);
+            try {
+                runnable.run();
+                return eventExecutor.newSucceededFuture(null);
+            } catch (Throwable t) {
+                return eventExecutor.newFailedFuture(t);
+            }
         }
```

Comment:
This change is the one that scares me the most, but only because it might uncover some bugs we have elsewhere.

Reply:
I've changed the usages of `doInEventLoop` that return a future from within the runnable that's passed in fail the future in case future returned by `doInEventLoop` fails.